feat: Redirects Territory (Town) Entering/Leaving Messages to NotificationManager Overlay

san7890 opened the pull request:
![image](https://user-images.githubusercontent.com/34697715/204661273-27909819-154c-47a2-abd1-b288f3f22ac9.png)

More trimming down on having these messages (that get old fast) being in the subtitle. Since there was both a subtitle and a chat message that said the exact same thing, I put it in it's own file to handle both instances of it showing up- let me know if I should just roll it elsewhere. I figure this way is probably better so you only have to edit one regex should Wynncraft change how it sends these messages and don't have to worry about desync.

Also, let me know if there are any recommendations for altering the formatting of the message itself. I went through about five different variations and this was the only one that I liked, will appreciate feedback on making it a bit better if it looks ugly.

san7890 committed:
Latticework

san7890 committed:
simplifies string, formatting

san7890 committed:
spotless

san7890 committed:
town -> territory

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,53 @@
+        String enteringMessage = String.format("§7Now §o%s§r %s", direction, territoryName);
+        NotificationManager.queueMessage(enteringMessage);
+    }
+    // Handles the chat log message event, we don't want a duplicate so just cancel the event and rely on the subtitle
```
```suggestion
    
    // Handles the chat log message event, we don't want a duplicate so just cancel the event and rely on the subtitle
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,53 @@
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class TerritoryMessageRedirectFeature extends UserFeature {
```
Again, why did we not fold this into InfoMessageFilter?

kristofbolyai replied:
Oh! This is a subtitle event. Okay. Then it is fine to leave it as is. For now. Maybe, we should create an universal feature for titles, like InfoMessageFilter feature is for chat. Not scope of this PR though.

kristofbolyai replied:
Add the \@FeatureInfo annotation though.

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,53 @@
+        // message (looks odd otherwise).
+        String territoryName = rawTerritoryName.substring(0, 1).toUpperCase() + rawTerritoryName.substring(1);
+
+        String enteringMessage = String.format("§7Now §o%s§r %s", direction, territoryName);
```
I think we could shorten this with using an arrow emoji, like friend join message does.

san7890 replied:
See PR Body for new example with emoji. I think the italicization still "works", but let me know if you want me to axe it.

kristofbolyai replied:
I would just straight up remove the entering/leaving part. No need for it. Arrows are condense and easy to understand. I would however, make the whole message gray, territory name included. 

As much as you like your screen/chat not the be spammed with random stuff, I like it when notifications are short and simple to understand.

san7890 replied:
Updated, will show up in coming commits.

san7890 replied:
PR Body Image updated

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,53 @@
+    public void onSubtitleSetText(SubtitleSetTextEvent event) {
+        if (!redirectTerritoryMessages) {
+            return;
+        }
```
```suggestion
        if (!redirectTerritoryMessages) return;
        
```

san7890 committed:
Formatting Code Review

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

san7890 committed:
ci: spotless formatting

san7890 committed:
uses arrow emoji

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,66 @@
+    private static final Pattern TERRITORY_MESSAGE_PATTERN = Pattern.compile("§7\\[You are now (\\S+) (.+)\\]");
+
+    @Config
+    public boolean redirectTerritoryMessages = true;
```
This config is useless. If the user turns off this feature, they get the same behaviour. Remove it.

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -579,6 +579,9 @@
   "feature.wynntils.territoryDefenseMessage.name": "Territory Defense Messages",
+  "feature.wynntils.territoryMessageRedirect.name": "Territory Message Redirect",
+  "feature.wynntils.territoryMessageRedirect.redirectTerritoryMessages.description": "Should messages about entering or leaving a Territory be redirected to the game update ticker?",
+  "feature.wynntils.territoryMessageRedirect.redirectTerritoryMessages.name": "Redirect Entering/Leaving Territory Messages",
```
Don't forget to remove the i18n as well

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,66 @@
+
+    // Handles the chat log message event, we don't want a duplicate so just cancel the event and rely on the subtitle
+    // text event.
+    @SubscribeEvent(priority = EventPriority.LOW)
```
This is fine to be a normal event.
```suggestion
    @SubscribeEvent
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,66 @@
+        if (!redirectTerritoryMessages) {
+            return;
+        }
+        String codedString = ComponentUtils.getCoded(event.getMessage());
```
No need for this. Use `event.getOriginalCodedMessage()`

san7890 committed:
removes entering/leaving, arrow only

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -579,6 +579,9 @@
   "feature.wynntils.statusOverlay.overlay.status.textShadow.description": "What should the text shadow look like?",
   "feature.wynntils.statusOverlay.overlay.status.textShadow.name": "Text Shadow",
   "feature.wynntils.territoryDefenseMessage.name": "Territory Defense Messages",
+  "feature.wynntils.territoryMessageRedirect.name": "Territory Message Redirect",
```
```suggestion
  "feature.wynntils.territoryMessageRedirect.name": "Territory Message Redirects",
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,66 @@
+
+        String codedString = ComponentUtils.getCoded(event.getComponent());
+        Matcher matcher = TERRITORY_MESSAGE_PATTERN.matcher(codedString);
+        if (matcher.matches()) event.setCanceled(true);
```
```suggestion
        if (matcher.matches()) event.setCanceled(true);
        
```

kristofbolyai replied:
Also, didn't you mean to return early on the case this does not match? And cancel the event otherwise?

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,66 @@
+                directionalArrow = "←";
+                break;
+        }
+        // Want to account for weird stuff like "the Forgery" and make it "The Forgery" for the sake of our brief
```
```suggestion

        // Want to account for weird stuff like "the Forgery" and make it "The Forgery" for the sake of our brief
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,66 @@
+        String rawDirection = matcher.group(1);
+        String rawTerritoryName = matcher.group(2);
+        String directionalArrow = "";
+        switch (rawDirection) {
```
```suggestion

        switch (rawDirection) {
```

kristofbolyai reviewed (commented):
Leaving new lines between different parts of a method can go a long way in making the method look much cleaner!

san7890 committed:
Cleanout

Removes Config, Priority, makes it just getOriginalCodedMessage, linebreaks, l18n cleanup.

kristofbolyai commented on common/src/main/java/com/wynntils/core/features/FeatureRegistry.java:
```diff
@@ -178,6 +179,7 @@ public static void init() {
         registerFeature(new PlayerArmorHidingFeature());
         registerFeature(new PlayerGhostTransparencyFeature());
         registerFeature(new InventoryRedirectFeature());
+        registerFeature(new TerritoryMessageRedirectFeature());
```
These 2 are not in an alphabetical order.

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,58 @@
+    // Handles the chat log message event, we don't want a duplicate so just cancel the event and rely on the subtitle
+    // text event.
+    @SubscribeEvent
+    public void onChat(ChatMessageReceivedEvent event) {
```
The whole method could be just "if( TERRPATTERN.matcher(event.get..()).matches()) cancel event"

san7890 committed:
alphabetization

san7890 committed:
Flattens ChatMessage into one line

and comment cleanup

san7890 committed:
adds default/return to swtich statement

san7890 committed:
DO YOU WANT ME TO INITIALIZE IT OR NOT

san7890 committed:
early return

san7890 committed:
ci: spotless formatting

magicus commented on common/src/main/java/com/wynntils/features/user/redirects/TerritoryMessageRedirectFeature.java:
```diff
@@ -0,0 +1,60 @@
+
+        // Want to account for weird stuff like "the Forgery" and make it "The Forgery"
+        // for the sake of our brief message (looks odd otherwise).
+        String territoryName = rawTerritoryName.substring(0, 1).toUpperCase() + rawTerritoryName.substring(1);
```
Is there not already a utility function to do this? Either in our classes or like google CaseFormat. If not, consider making this a string util function.

san7890 replied:
it was in apache stringutils, whoops

san7890 committed:
swaps capitalization to StringUtils

san7890 committed:
Merge remote-tracking branch 'upstream/alpha' into weird-acronyms

kristofbolyai commented:
Thank you for your contribution !
